### PR TITLE
#157 Added new rule for nested if statements.

### PR DESCRIPTION
**Pull request by iwarapter**

### CODE REVIEW

File: puppet-checks/src/test/groovy/com/iadams/sonarqube/puppet/checks/NestedIfStatementsCheckSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * SOFTWARE.
+ */
+package com.iadams.sonarqube.puppet.checks
+
+import com.iadams.sonarqube.puppet.PuppetAstScanner
+import org.sonar.squidbridge.api.SourceFile
+import org.sonar.squidbridge.checks.CheckMessagesVerifier
+import spock.lang.Specification
+
+class NestedIfStatementsCheckSpec extends Specification {
```
Comment (racodond) [inline puppet-checks/src/test/groovy/com/iadams/sonarqube/puppet/checks/NestedIfStatementsCheckSpec.groovy:32]: @iwarapter Could you please add test cases with `elseif` statements and check that the rule behaves as expected?

File: puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/NestedIfStatementsCheck.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import org.sonar.squidbridge.checks.SquidCheck;
+
+@Rule(
+  key = "NestedIfStatements",
+  name = "If statements should not be nested too deeply",
+  priority = Priority.MAJOR,
+  tags = Tags.BRAIN_OVERLOAD
+)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_CHANGEABILITY)
+@SqaleConstantRemediation("10min")
```
Comment (racodond) [inline puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/NestedIfStatementsCheck.java:46]: @iwarapter You should use the SQALE linear remediation function or the linear remediation function with offset. See http://www.sonarsource.com/products/plugins/governance/sqale/installation-and-usage/#customizingAnalysisModel for more details. You can mimic what I did in https://github.com/iwarapter/sonar-puppet/commit/412b435f0d6569deb3c46079662b3c0cc72bc5e3. By the way, we should now use the new `PuppetCheckVisitor` I created.

File: puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/NestedIfStatementsCheck.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+import org.sonar.squidbridge.checks.SquidCheck;
+
+@Rule(
+  key = "NestedIfStatements",
+  name = "If statements should not be nested too deeply",
```
Comment (racodond) [inline puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/NestedIfStatementsCheck.java]: "\"if\"" statements should...

File: puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/NestedIfStatementsCheck.java (hunk 2 above)
Comment (iwarapter) [inline puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/NestedIfStatementsCheck.java:46]: @racodond the problem with this is this rule [(and all the other language rules)](http://nemo.sonarqube.org/coding_rules#q=control%20statements%20should%20not%20nest) only flag on the first instance of it being broken not for each issue. We should keep this rule inline with the sonar standards.

Comment (racodond) [inline puppet-checks/src/main/java/com/iadams/sonarqube/puppet/checks/NestedIfStatementsCheck.java:46]: @iwarapter OK

**Comment by racodond** [comment]
@iwarapter Could you also add some tests with if statements nested in else statements to make sure it works fine as well?

**Comment by iwarapter** [comment]
@racodond like [this one](https://github.com/iwarapter/sonar-puppet/commit/e7ce653fed3767fbf9fe90be8878e983f5a32f32#diff-c1d5af2b8ad6c5bb1c677caaf1a2c6faR3)?

**Comment by racodond** [comment]
Missed it. Perfect. Thanks!

**Pull request merged**